src/cgmemmgr.cpp

```diff
@@ -0,0 +1,839 @@
+                                 unsigned SectionID, StringRef SectionName,
+                                 bool isReadOnly) override;
+#ifdef LLVM38
+    void notifyObjectLoaded(RuntimeDyld &Dyld,
+                            const object::ObjectFile &Obj) override;
+#endif
+    bool finalizeMemory(std::string *ErrMsg = nullptr) override;
+    template <typename DL, typename Alloc>
+    void mapAddresses(DL &Dyld, Alloc *allocator)
+    {
+        for (auto &alloc: allocator->allocations) {
+            if (alloc.rt_addr == alloc.wr_addr || alloc.relocated)
+                continue;
+            alloc.relocated = true;
+            Dyld.mapSectionAddress(alloc.wr_addr, (uintptr_t)alloc.rt_addr);
+        }
+    }
+    template <typename DL>
+    void mapAddresses(DL &Dyld)
+    {
+        if (!ro_alloc)
+            return;
+        mapAddresses(Dyld, ro_alloc);
+        mapAddresses(Dyld, exe_alloc);
+    }
+#ifdef _OS_WINDOWS_
+    template <typename Alloc>
+    void *lookupWriteAddressFor(void *rt_addr, Alloc *allocator)
+    {
+        for (auto &alloc: allocator->allocations) {
+            if (alloc.rt_addr == rt_addr) {
+                return alloc.wr_addr;
+            }
+        }
+        return nullptr;
+    }
+    void *lookupWriteAddressFor(void *rt_addr)
+    {
+        if (!ro_alloc)
+            return rt_addr;
+        if (void *ptr = lookupWriteAddressFor(rt_addr, ro_alloc))
+            return ptr;
+        if (void *ptr = lookupWriteAddressFor(rt_addr, exe_alloc))
+            return ptr;
+        return rt_addr;
+    }
+#endif // _OS_WINDOWS_
+};
+
+uint8_t *RTDyldMemoryManagerJL::allocateCodeSection(uintptr_t Size,
+                                                    unsigned Alignment,
+                                                    unsigned SectionID,
+                                                    StringRef SectionName)
+{
+    if (exe_alloc)
+        return (uint8_t*)exe_alloc->alloc(Size, Alignment);
+    return SectionMemoryManager::allocateCodeSection(Size, Alignment, SectionID,
+                                                     SectionName);
+}
+
+uint8_t *RTDyldMemoryManagerJL::allocateDataSection(uintptr_t Size,
+                                                    unsigned Alignment,
+                                                    unsigned SectionID,
+                                                    StringRef SectionName,
+                                                    bool isReadOnly)
+{
+    if (!isReadOnly)
+        return (uint8_t*)rw_alloc.alloc(Size, Alignment);
+    if (ro_alloc)
+        return (uint8_t*)ro_alloc->alloc(Size, Alignment);
+    return SectionMemoryManager::allocateDataSection(Size, Alignment, SectionID,
+                                                     SectionName, isReadOnly);
+}
+
+#ifdef LLVM38
+void RTDyldMemoryManagerJL::notifyObjectLoaded(RuntimeDyld &Dyld,
+                                               const object::ObjectFile &Obj)
```

Comment:
To be fair, I think I actually had override here initially without realizing that it is only allowed in the declaration......